src/dtype.rs

```diff
@@ -217,74 +155,142 @@ impl DataType {
-    #[inline]
-    fn npy_type() -> NPY_TYPES {
-        Self::DATA_TYPE.into_ctype()
+fn npy_int_type_lookup<T, T0, T1, T2>(npy_types: [NPY_TYPES; 3]) -> NPY_TYPES {
+    // `npy_common.h` defines the integer aliases. In order, it checks:
+    // NPY_BITSOF_LONG, NPY_BITSOF_LONGLONG, NPY_BITSOF_INT, NPY_BITSOF_SHORT, NPY_BITSOF_CHAR
+    // and assigns the alias to the first matching size, so we should check in this order.
+    match size_of::<T>() {
+        x if x == size_of::<T0>() => npy_types[0],
+        x if x == size_of::<T1>() => npy_types[1],
+        x if x == size_of::<T2>() => npy_types[2],
+        _ => panic!("Unable to match integer type descriptor: {:?}", npy_types),
     }
+}
 
-    /// Create `dtype`.
-    fn get_dtype(py: Python) -> &PyArrayDescr {
-        PyArrayDescr::from_npy_type(py, Self::npy_type())
+fn npy_int_type<T: Bounded + Zero + Sized + PartialEq>() -> NPY_TYPES {
+    let is_unsigned = T::min_value() == T::zero();
+    let bit_width = size_of::<T>() << 3;
+
+    match (is_unsigned, bit_width) {
+        (false, 8) => NPY_TYPES::NPY_BYTE,
+        (false, 16) => NPY_TYPES::NPY_SHORT,
+        (false, 32) => npy_int_type_lookup::<i32, c_long, c_int, c_short>([
+            NPY_TYPES::NPY_LONG,
+            NPY_TYPES::NPY_INT,
+            NPY_TYPES::NPY_SHORT,
+        ]),
+        (false, 64) => npy_int_type_lookup::<i64, c_long, c_longlong, c_int>([
+            NPY_TYPES::NPY_LONG,
+            NPY_TYPES::NPY_LONGLONG,
+            NPY_TYPES::NPY_INT,
+        ]),
+        (true, 8) => NPY_TYPES::NPY_UBYTE,
+        (true, 16) => NPY_TYPES::NPY_USHORT,
+        (true, 32) => npy_int_type_lookup::<u32, c_ulong, c_uint, c_ushort>([
+            NPY_TYPES::NPY_ULONG,
+            NPY_TYPES::NPY_UINT,
+            NPY_TYPES::NPY_USHORT,
+        ]),
+        (true, 64) => npy_int_type_lookup::<u64, c_ulong, c_ulonglong, c_uint>([
+            NPY_TYPES::NPY_ULONG,
+            NPY_TYPES::NPY_ULONGLONG,
+            NPY_TYPES::NPY_UINT,
+        ]),
+        _ => unreachable!(),
     }
 }
 
-macro_rules! impl_num_element {
-    ($t:ty, $npy_dat_t:ident $(,$npy_types: ident)+) => {
-        unsafe impl Element for $t {
-            const DATA_TYPE: DataType = DataType::$npy_dat_t;
-            fn is_same_type(dtype: &PyArrayDescr) -> bool {
-                $(dtype.get_typenum() == NPY_TYPES::$npy_types as i32 ||)+ false
+macro_rules! impl_element_scalar {
+    (@impl: $ty:ty, $npy_type:expr $(,#[$meta:meta])*) => {
+        $(#[$meta])*
+        unsafe impl Element for $ty {
+            const IS_COPY: bool = true;
+            fn get_dtype(py: Python) -> &PyArrayDescr {
+                PyArrayDescr::from_npy_type(py, $npy_type)
             }
         }
     };
+    ($ty:ty => $npy_type:ident $(,#[$meta:meta])*) => {
+        impl_element_scalar!(@impl: $ty, NPY_TYPES::$npy_type $(,#[$meta])*);
+    };
+    ($ty:ty) => {
```

Comment:
Couldn't these two arm be merged via using `$($tys:ty),*` below?

Reply:
Good catch, missed that.